### DISCUSSION TITLE
What's the correct way to create a library that works in a monorepo, has an example app, and exposes its App.cpp file?

### DISCUSSION THREAD

**Discussion by shirakaba** [microsoft/react-native-windows#13560] [Q&A] [answered]
I've spent the last week trying out everything I can think of to just develop a library in my monorepo, following the [native module setup](https://microsoft.github.io/react-native-windows/docs/native-modules-setup#creating-a-blank-native-module-project) docs and more. I've:

1. Used `react-native-windows-init`:
    - … with `experimentalNuGetDependency` both enabled and disabled
    - … with both `app` and `lib` projectTypes
2. Forked [React Native Document Picker](https://github.com/react-native-documents/document-picker#readme)
3. Forked [@react-native-web-apis/web-storage](https://github.com/microsoft/rnx-kit/tree/main/incubator/%40react-native-webapis/web-storage)
4. Forked [react-native-randombytes](https://github.com/mvayngrib/react-native-randombytes/tree/master)
5. Forked [react-native-svg](https://github.com/software-mansion/react-native-svg)
6. Forked [test-app](https://github.com/microsoft/rnx-kit/tree/main/packages/test-app)
7. Forked the `react-native-test-app` [example](https://github.com/microsoft/react-native-test-app/tree/582c67cf24b72a2b38391d1a8784754e6ded3543/example)
8. Gave up on building an example app and tried just consuming the native module directly from my main app
9. Gave up on building the libraries inside the monorepo and tried working on packages in a regular repo instead.

For one reason or another, I've not been able to get *any* of these working. It's genuinely quite exhausting as the builds take forever and the errors are impenetrable.

My environment:
- I'm running on an ARM Mac through an ARM Windows VM, building for ARM.
- I'm using a pnpm monorepo.
- I'm targeting `react-native-windows` v0.73 to align with `react-native-macos`.
 
Reasons for failed attempts include:
  - Projects turn out to be library-only, so lack boilerplate for an example app.
  - [react-native-test-app](https://github.com/microsoft/react-native-test-app) projects don't expose the `App.cpp` file (this is a hard requirement, because my native module involves deeplinking).
  - The `windows` folder is empty and running the commands provided in the boilerplate doesn't populate it
  - The `.sln` file keeps being generated with the wrong paths to files, due to working in a monorepo that hoists dependencies.
  - The project initialisation [fails when using an org prefix](https://github.com/microsoft/react-native-windows/issues/13558).
  - The conversion of project name to camel-case in one tool (e.g. `ReactNativeAppAuthJS` -> `ReactNativeAppAuthJs`) breaks other downstream tools that neglect to do the same conversion, so the `.sln` file can't be found.
  - I get `Command failed with error MSBuildError: C:\Users\jamie\git\my_monorepo\apps\noman\node_modules\react-native-app-auth-js\windows\Generated Files\ReactPackageProvider.g.cpp(5,49): error C2039: 'ReactNativeAppAuthJs': is not a member of 'winrt'` (when forking [@react-native-web-apis/web-storage](https://github.com/microsoft/rnx-kit/tree/main/incubator/%40react-native-webapis/web-storage)).
  - [Not building for ARM](https://github.com/microsoft/react-native-windows/issues/13559) (when using `react-native-windows-init --projectType=app --experimentalNuGetDependency`).
  - Ending up with [two copies of `hermes.dll`](https://github.com/microsoft/react-native-windows/issues/13559#issuecomment-2282709950) (when using `react-native-windows-init --projectType=app`).

I'd love to get involved with contributing Windows support to React Native packages, but I can't even start the first step. Is there any foolproof workflow for simply creating a native module, with an example app, which exposes its `App.cpp` file, in a `pnpm` monorepo?

**Comment by shirakaba** [comment]
Note to self – although I had no luck with `react-native-windows-init`, I still need to try out the newer `npx react-native init-windows` command as suggested here: https://github.com/microsoft/react-native-windows/issues/13558#issuecomment-2284697890

**Reply by chiaramooney** [reply]
Tagging @jonthysell who worked on newer init command. Jon, do you have input on this?

**Comment by chrisglein** [comment, marked as answer]
Linking related issue there: https://github.com/microsoft/react-native-windows/issues/13559

Summary is that this is pretty far off of our golden path (and will have lower priority as we focus on improving the golden path for the new architecture). That doesn't mean this will be impossible, but there are going to be rough edges and individual issues that need to be resolved one by one. We'll track those as issues logged on the repo.
